module/state_synchronization/indexer/collection_executed_metric.go

```diff
@@ -87,10 +93,25 @@ func (c *CollectionExecutedMetricImpl) BlockFinalized(block *flow.Block) {
 		}
 
 		for _, t := range l.Transactions {
+			blockTransactions.transactions = append(blockTransactions.transactions, t)
 			c.accessMetrics.TransactionFinalized(t, now)
 		}
 	}
 
+	for _, s := range block.Payload.Seals {
+		_, err := c.blocks.ByID(s.BlockID)
+		if err != nil {
+			c.log.Warn().Err(err).Msg("could not find block")
+			continue
+		}
+
+		if s.BlockID == blockTransactions.blockID && len(blockTransactions.transactions) != 0 {
+			for _, t := range blockTransactions.transactions {
```

Comment:
I don't think this would work. 

The txs from `blockTransactions.transactions` are transactions from this block, but the block.PayloadSeals usually is for other blocks. This logic only works if the finalized block is also the sealed block, which is rarely the case. 

Since finalization and sealing are async (meaning a finalized block is sealed by another block), I think we need to use a mempool to keep track of finalized transactions, and remove them when they are sealed.

Reply:
yea, I think you will need to store `blockTransactions` on the `CollectionExecutedMetricImpl` object so it's reused across blocks. then you can remove the entry for the block after checking here like Leo said.